src/lib.rs

```diff
@@ -553,231 +561,45 @@ impl State {
-    wakers: Vec<(usize, Waker)>,
+    /// Run a tick using the provided function to get the next task.
+    async fn tick_with(&self, mut local_ticker: impl FnMut(bool) -> Option<Runnable>) -> Runnable {
+        let local = self.local_queue.get_or_default();
 
-    /// Reclaimed IDs.
-    free_ids: Vec<usize>,
-}
-
-impl Sleepers {
-    /// Inserts a new sleeping ticker.
-    fn insert(&mut self, waker: &Waker) -> usize {
-        let id = match self.free_ids.pop() {
-            Some(id) => id,
-            None => self.count + 1,
-        };
-        self.count += 1;
-        self.wakers.push((id, waker.clone()));
-        id
-    }
-
-    /// Re-inserts a sleeping ticker's waker if it was notified.
-    ///
-    /// Returns `true` if the ticker was notified.
-    fn update(&mut self, id: usize, waker: &Waker) -> bool {
-        for item in &mut self.wakers {
-            if item.0 == id {
-                if !item.1.will_wake(waker) {
-                    item.1 = waker.clone();
-                }
-                return false;
+        loop {
+            // Try to get a runnable from the local queue.
+            if let Some(runnable) = local_ticker(false) {
+                return runnable;
             }
-        }
 
-        self.wakers.push((id, waker.clone()));
-        true
-    }
+            // Register a local waiter.
+            listener!(local.waiters => local_listener);
 
-    /// Removes a previously inserted sleeping ticker.
-    ///
-    /// Returns `true` if the ticker was notified.
-    fn remove(&mut self, id: usize) -> bool {
-        self.count -= 1;
-        self.free_ids.push(id);
-
-        for i in (0..self.wakers.len()).rev() {
-            if self.wakers[i].0 == id {
-                self.wakers.remove(i);
-                return false;
+            // Try for a global runner.
+            if let Ok(runnable) = self.queue.pop() {
+                return runnable;
             }
-        }
-        true
-    }
 
-    /// Returns `true` if a sleeping ticker is notified or no tickers are sleeping.
-    fn is_notified(&self) -> bool {
-        self.count == 0 || self.count > self.wakers.len()
-    }
-
-    /// Returns notification waker for a sleeping ticker.
-    ///
-    /// If a ticker was notified already or there are no tickers, `None` will be returned.
-    fn notify(&mut self) -> Option<Waker> {
-        if self.wakers.len() == self.count {
-            self.wakers.pop().map(|item| item.1)
-        } else {
-            None
-        }
-    }
-}
-
-/// Runs task one by one.
-struct Ticker<'a> {
-    /// The executor state.
-    state: &'a State,
-
-    /// Set to a non-zero sleeper ID when in sleeping state.
-    ///
-    /// States a ticker can be in:
-    /// 1) Woken.
-    /// 2a) Sleeping and unnotified.
-    /// 2b) Sleeping and notified.
-    sleeping: usize,
-}
+            // Register a global waiter.
+            listener!(self.new_tasks => global_listener);
```

Comment:
Isn't this too late to register the global waiter?

Reply:
No, since we check the global queue again later.